gpbft/participant.go

```diff
@@ -123,41 +114,60 @@ func (p *Participant) ReceiveMessage(msg *GMessage) (bool, error) {
 		if err := p.granite.Receive(msg); err != nil {
 			return true, fmt.Errorf("receiving message: %w", err)
 		}
-		p.handleDecision()
-		return true, nil
+		err = p.handleDecision()
+		return true, err
 	} else if msg.Vote.Instance >= p.nextInstance {
 		// Queue messages for later instances
 		return false, xerrors.Errorf("message for future instance cannot be valid")
 	}
 	// Message dropped.
-	return false, err
+	return false, nil
 }
 
-func (p *Participant) ReceiveAlarm() error {
-	var err error
+func (p *Participant) ReceiveAlarm() (err error) {
 	defer func() {
 		if r := recover(); r != nil {
 			err = &PanicError{Err: err}
 		}
 	}()
 
-	if p.granite != nil {
-		// An instance is robust to receiving extra alarms, e.g. from prior terminated instances.
+	if p.granite == nil {
+		// The alarm is for fetching the next chain and beginning a new instance.
+		return p.beginInstance()
+	} else {
 		if err := p.granite.ReceiveAlarm(); err != nil {
 			return fmt.Errorf("failed receiving alarm: %w", err)
 		}
-		p.handleDecision()
+		return p.handleDecision()
 	}
-	return err
 }
 
-func (p *Participant) handleDecision() {
+func (p *Participant) beginInstance() error {
+	chain, power, beacon := p.host.GetCanonicalChain(p.nextInstance)
+	var err error
+	if p.granite, err = newInstance(p.config, p.host, p.id, p.nextInstance, chain, power, beacon); err != nil {
+		return fmt.Errorf("failed creating new granite instance: %w", err)
+	}
+	p.nextInstance += 1
+	if err := p.granite.Start(); err != nil {
+		return fmt.Errorf("failed starting granite instance: %w", err)
+	}
+	return p.handleDecision()
+}
+
+func (p *Participant) handleDecision() error {
 	if p.terminated() {
 		p.finalised = p.granite.terminationValue
 		p.terminatedDuringRound = p.granite.round
 		p.granite = nil
 		p.host.ReceiveDecision(*p.finalised)
+
+		// Set an alarm at which to fetch the next chain and begin a new instance.
+		// TODO: replace this zero delay with something based on tipset timestamps.
+		p.host.SetAlarm(p.host.Time())
```

Comment:
Ah well, this is only true if we start a new instance only once our head locally changes. Since we start instance following an alarm, we should probably stick to the alarm not to have many instances finalizing baseChain needlessly (during an epoch with no proposals, for example).

Reply:
I will replace this TODO with an issue before merging. As you note, there's no point starting a new instance if there's no new chain to propose.